### PR TITLE
IOperation analyzers

### PR DESCRIPTION
**Pull request by sharwell**
Rewrite analyzers as IOperation analyzers to improve performance. This pull request changes the three slowest analyzers according to https://github.com/dotnet/roslyn/pull/43394#issuecomment-620776861.

### CODE REVIEW

File: src/Microsoft.VisualStudio.Threading.Analyzers.CodeFixes/VSTHRD103UseAsyncOptionCodeFix.cs [hunk 1]
```diff
@@ -135,13 +135,16 @@ protected override async Task<Solution> GetChangedSolutionAsync(CancellationToke
                 var originalAnonymousMethodContainerIfApplicable = syncMethodName.FirstAncestorOrSelf<AnonymousFunctionExpressionSyntax>();
                 var originalMethodDeclaration = syncMethodName.FirstAncestorOrSelf<MethodDeclarationSyntax>();
 
+                var enclosingSymbol = semanticModel.GetEnclosingSymbol(this.diagnostic.Location.SourceSpan.Start, cancellationToken);
+                var hasReturnValue = ((enclosingSymbol as IMethodSymbol)?.ReturnType as INamedTypeSymbol)?.IsGenericType ?? false;
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.Threading.Analyzers.CodeFixes/VSTHRD103UseAsyncOptionCodeFix.cs:139]: 💭 This is the one I was least sure of

File: src/Microsoft.VisualStudio.Threading.Analyzers.Tests/VSTHRD109AvoidAssertInAsyncMethodsAnalyzerTests.cs [hunk 2]
```diff
@@ -252,11 +247,10 @@ class Test {
 }
 ";
 
-            var expected = Verify.Diagnostic().WithSpan(9, 26, 9, 46);
-            await Verify.VerifyCodeFixAsync(test, expected, fix);
+            await Verify.VerifyCodeFixAsync(test, fix);
         }
 
-        [Fact(Skip = "Fails because the semantic model claims the anonymous func returns Task instead of Task<int>. We should probably skip that check and always preserve return statements.")]
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.Threading.Analyzers.Tests/VSTHRD109AvoidAssertInAsyncMethodsAnalyzerTests.cs:259]: 📝 This was caused by the use of a semantic model from an intermediate state - the method was not yet marked as `async` but an `await` expression had already been added.

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs [hunk 3]
```diff
@@ -65,27 +59,24 @@ public override void Initialize(AnalysisContext context)
...
-        /// </summary>
-        private void AnalyzeInvocation(SyntaxNodeAnalysisContext context)
+        private void AnalyzeInvocation(OperationAnalysisContext context)
         {
-            var invocation = (InvocationExpressionSyntax)context.Node;
-            var symbol = context.SemanticModel.GetSymbolInfo(invocation.Expression, context.CancellationToken).Symbol;
-            if (symbol != null)
+            var invocation = (IInvocationOperation)context.Operation;
+            var symbol = invocation.TargetMethod;
+            if (symbol is object)
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs:79]: symbol is guaranteed to be non-null

Comment (mavasani) [reply src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs:79]: also the symbol kind check below is redundant - it will always be a method.

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs [hunk 4]
```diff
@@ -49,43 +44,26 @@ public override void Initialize(AnalysisContext context)
...
+                ctxt.RegisterOperationAction(Utils.DebuggableWrapper(c => this.AnalyzeInvocation(c, mainThreadAssertingMethods)), OperationKind.Invocation);
             });
         }
 
-        private void AnalyzeInvocation(SyntaxNodeAnalysisContext context, ImmutableArray<CommonInterest.QualifiedMember> mainThreadAssertingMethods)
+        private void AnalyzeInvocation(OperationAnalysisContext context, ImmutableArray<CommonInterest.QualifiedMember> mainThreadAssertingMethods)
         {
-            var functionInfo = Utils.GetContainingFunction((CSharpSyntaxNode)context.Node);
-            if (functionInfo.Function == null)
+            if (!(Utils.GetContainingFunction(context.Operation, context.ContainingSymbol) is IMethodSymbol methodSymbol))
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs:53]: Why do we need to invoke `Utils.GetContainingFunction` here?
```suggestion
            if (!(context.ContainingSymbol is IMethodSymbol methodSymbol))
```

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs [hunk 5]
```diff
@@ -49,43 +44,26 @@ public override void Initialize(AnalysisContext context)
...
-            }
-
-            if (functionInfo.IsAsync || Utils.HasAsyncCompatibleReturnType(methodSymbol as IMethodSymbol) || Utils.IsAsyncCompatibleReturnType(implicitReturnType))
-            {
-                var invocationExpression = (InvocationExpressionSyntax)context.Node;
-                var symbolInfo = context.SemanticModel.GetSymbolInfo((InvocationExpressionSyntax)context.Node, context.CancellationToken);
-                var symbol = symbolInfo.Symbol;
+                var invocation = (IInvocationOperation)context.Operation;
+                var symbol = invocation.TargetMethod;
                 if (symbol != null)
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs:62]: Redundant null check

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD004AwaitSwitchToMainThreadAsyncAnalyzer.cs [hunk 6]
```diff
@@ -48,22 +36,21 @@ public override void Initialize(AnalysisContext context)
...
 
-        private void AnalyzeInvocation(SyntaxNodeAnalysisContext context)
+        private void AnalyzeInvocation(OperationAnalysisContext context)
         {
-            var invocation = (InvocationExpressionSyntax)context.Node;
-            var invokedSymbol = context.SemanticModel.GetSymbolInfo(invocation.Expression, context.CancellationToken);
-            if (invokedSymbol.Symbol is IMethodSymbol methodSymbol &&
+            var invocation = (IInvocationOperation)context.Operation;
+            if (invocation.TargetMethod is IMethodSymbol methodSymbol &&
                 methodSymbol.Name == Types.JoinableTaskFactory.SwitchToMainThreadAsync &&
```
Comment (mavasani) [inline src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD004AwaitSwitchToMainThreadAsyncAnalyzer.cs:46]:
```suggestion
            var methodSymbol = (IMethodSymbol)invocation.TargetMethod;
            if (methodSymbol.Name == Types.JoinableTaskFactory.SwitchToMainThreadAsync &&
```

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs [hunk 7]
```diff
@@ -65,27 +59,24 @@ public override void Initialize(AnalysisContext context)
             context.EnableConcurrentExecution();
             context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
 
-            context.RegisterCodeBlockStartAction<SyntaxKind>(ctxt =>
+            context.RegisterOperationBlockStartAction(context =>
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs:62]: Interesting. I didn't realize you could overlap an outer scoped variable with a lambda variable.

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs (hunk 3 above)
Comment (AArnott) [reply src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs:79]: It'll be great when this whole API is null ref annotated to make these guarantees more obvious. :)

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs (hunk 4 above)
Comment (AArnott) [reply src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs:53]: Because the containing symbol is always the method that is a member of the containing type. But `GetContainingFunction` returns the ... containing function, which might be an inline function, lambda, or delegate.

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs [hunk 8]
```diff
@@ -49,43 +44,26 @@ public override void Initialize(AnalysisContext context)
...
-        private void AnalyzeInvocation(SyntaxNodeAnalysisContext context, ImmutableArray<CommonInterest.QualifiedMember> mainThreadAssertingMethods)
+        private void AnalyzeInvocation(OperationAnalysisContext context, ImmutableArray<CommonInterest.QualifiedMember> mainThreadAssertingMethods)
         {
-            var functionInfo = Utils.GetContainingFunction((CSharpSyntaxNode)context.Node);
-            if (functionInfo.Function == null)
+            if (!(Utils.GetContainingFunction(context.Operation, context.ContainingSymbol) is IMethodSymbol methodSymbol))
             {
-                // One case where this happens is the use of nameof(X) in the argument of a custom attribute on a type, such as:
-                // [System.Diagnostics.DebuggerDisplay(""hi"", Name = nameof(System.Console))]
-                // class Foo { }
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs:63]: Why did you remove the explanatory code comment?

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs (hunk 7 above)
Comment (sharwell) [reply src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs:62]: New for C# 8.

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs (hunk 4 above)
Comment (AArnott) [reply src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs:53]: At least, I _think_ that's the case.

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs (hunk 8 above)
Comment (sharwell) [reply src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs:63]: The comment is no longer correct. This is an `INameOfOperation`, not an invocation.

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs (hunk 3 above)
Comment (sharwell) [reply src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD106UseInvokeAsyncForAsyncEventsAnalyzer.cs]: Removed the null check

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs (hunk 5 above)
Comment (sharwell) [reply src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs]: Removed the null check

File: src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD004AwaitSwitchToMainThreadAsyncAnalyzer.cs (hunk 6 above)
Comment (sharwell) [reply src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD004AwaitSwitchToMainThreadAsyncAnalyzer.cs]: Now simplified

File: src/Microsoft.VisualStudio.Threading.Analyzers/Utils.cs [hunk 9]
```diff
@@ -225,6 +225,18 @@ internal static ExpressionSyntax IsolateMethodName(InvocationExpressionSyntax in
             return invokedMethodName;
         }
 
+        internal static SyntaxNode IsolateMethodName(IInvocationOperation invocation)
+        {
+            _ = invocation ?? throw new ArgumentNullException(nameof(invocation));
+
+            if (invocation.Syntax is InvocationExpressionSyntax { Expression: MemberAccessExpressionSyntax memberAccessExpression } invocationExpression)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Microsoft.VisualStudio.Threading.Analyzers/Utils.cs]: Done

File: src/Microsoft.VisualStudio.Threading.Analyzers/Utils.cs [hunk 10]
```diff
@@ -225,6 +225,18 @@ internal static ExpressionSyntax IsolateMethodName(InvocationExpressionSyntax in
             return invokedMethodName;
         }
 
+        internal static SyntaxNode IsolateMethodName(IInvocationOperation invocation)
+        {
+            _ = invocation ?? throw new ArgumentNullException(nameof(invocation));
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Microsoft.VisualStudio.Threading.Analyzers/Utils.cs]: Removed the null check

File: src/Microsoft.VisualStudio.Threading.Analyzers/Utils.cs [hunk 11]
```diff
@@ -371,6 +383,27 @@ internal static ContainingFunctionData GetContainingFunction(CSharpSyntaxNode sy
             return default(ContainingFunctionData);
         }
 
+        internal static ISymbol GetContainingFunction(IOperation operation, ISymbol operationBlockContainingSymbol)
+        {
+            while (operation is object)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Microsoft.VisualStudio.Threading.Analyzers/Utils.cs]: Rewrote the method using a temporary local
